### PR TITLE
Add AlertRule type and CRUD method

### PR DESCRIPTION
**Pull request by stevenbrichards**
Intending to add this for terraform support.

--

Based this on the example data examples in the API spec, so open to better test JSON / examples.

### CODE REVIEW

File: alerting_alert_rule.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"fmt"
+	"time"
+)
+
+// AlertRule represents a Grafana Alerting Alert Rule.
+type AlertRule struct {
+	Annotations  map[string]string `json:"annotations,omitempty"`
+	Conditions   string            `json:"conditions"`
+	Data         []*AlertQuery     `json:"data"`
+	ExecErrState string            `json:"execErrState"`
```
Comment (JohnnyQQQQ) [inline alerting_alert_rule.go]: This should be its own string type so we have some type-safety when using the client. Something like:

```golang
type ExecErrState string

ErrOK ExecErrState = "OK"
ErrError ExecErrState = "Error"
ErrAlerting ExecErrState = "Alerting"
```

**Review by JohnnyQQQQ — CHANGES_REQUESTED** [review]
Hey Steven 👋 thanks a lot for your contribution, I left a few minor comments.

File: alerting_alert_rule.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	FolderUID    string            `json:"folderUid"`
+	ID           int64             `json:"id,omitempty"`
+	Labels       map[string]string `json:"labels,omitempty"`
+	NoDataState  string            `json:"noDataState"`
```
Comment (JohnnyQQQQ) [inline alerting_alert_rule.go]: This should also be its own string type.

File: alerting_alert_rule.go (hunk 1 above)
Comment (stevenbrichards) [reply alerting_alert_rule.go]: ack, will do, thanks :-)

File: alerting_alert_rule.go (hunk 2 above)
Comment (stevenbrichards) [reply alerting_alert_rule.go]: fixed

**Comment by stevenbrichards** [comment]
@JohnnyQQQQ ready for re-review

**Comment by JohnnyQQQQ** [comment]
Thanks, the team member implementing this is currently on PTO and will take it from here as soon as he is back.

**Comment by stevenbrichards** [comment]
Happy to implement / contribute more if pointed in the direction of need, just let me know :-) We're keen to codify our alarms in terraform.

**Comment by alexweav** [comment]
Hi Steven, I'm actively working on this and the terraform integration for all alerting resources 👋 

For Terraform to work efficiently and be consistent with Prometheus, we need the ability to query all rules in a group. There's an open PR in the main grafana repo adding an endpoint for this. It's awaiting review: https://github.com/grafana/grafana/pull/51398

Once that PR goes in, we will then need a method here wrapping the new endpoint, then the terraform integration will be unblocked and good to go.

**Comment by stevenbrichards** [comment]
Hey Alex, thanks for the update. While implementing this I noticed that `POST /api/v1/provisioning/folder/{FolderUID}/rule-groups/{Group}` is also missing from the client, but is in the API docs. I'd be glad to implement the GET/POST methods for this resource with your thumbs-up, in anticipation of the PR referenced being merged.

File: alerting_alert_rule.go [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	ExecErrState ExecErrState      `json:"execErrState"`
+	FolderUID    string            `json:"folderUid"`
+	ID           int64             `json:"id,omitempty"`
+	Labels       map[string]string `json:"labels,omitempty"`
+	NoDataState  NoDataState       `json:"noDataState"`
+	OrgID        int64             `json:"orgId"`
+	RuleGroup    string            `json:"ruleGroup"`
+	Title        string            `json:"title"`
+	UID          string            `json:"uid,omitempty"`
+	Updated      string            `json:"updated,omitempty"`
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
	Updated      time.Time            `json:"updated"`
```

File: alerting_alert_rule.go [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+// AlertRule represents a Grafana Alerting Alert Rule.
+type AlertRule struct {
+	Annotations  map[string]string `json:"annotations,omitempty"`
+	Conditions   string            `json:"conditions"`
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
	Condition   string            `json:"condition"`
```

File: alerting_alert_rule.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
+package gapi
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"time"
+)
+
+// AlertRule represents a Grafana Alerting Alert Rule.
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
// AlertRule represents a Grafana Alert Rule.
```
Nitpick. This is inconsistent with the other comments but reads a bit better IMO.

File: alerting_alert_rule.go [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+	To   time.Duration `json:"to"`
+}
+
+// AlertRule fetches a single alert rule, identified by its UID.
+func (c *Client) AlertRule(uid string) (AlertRule, error) {
+	path := fmt.Sprintf("/api/v1/provisioning/alert-rules/%s", uid)
+	result := AlertRule{}
+	err := c.request("GET", path, nil, nil, &result)
+	if err != nil {
+		return AlertRule{}, fmt.Errorf("alert rule with uid %s not found", uid)
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
		return AlertRule{}, err
```
If it's a 404, the `err` will already say that the resource wasn't found in the message. For cases where the error isn't a 404, (unauthorized, can't reach server, etc) we don't want to return a potentially misleading message.

File: alerting_alert_rule.go [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+	ForDuration  time.Duration     `json:"for"`
+	Provenance   string            `json:"provenance"`
+}
+
+type AlertQuery struct {
+	DatasourceUID     string
+	Model             interface{}
+	QueryType         string
+	RefID             string
+	RelativeTimeRange RelativeTimeRange
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
	DatasourceUID     string `json:"datasourceUid,omitempty"`
	Model             interface{} `json:"model"`
	QueryType         string `json:"queryType,omitempty"`
	RefID             string `json:"refId,omitempty"`
	RelativeTimeRange RelativeTimeRange `json:"relativeTimeRange"`
```
Just for consistency.

File: alerting_alert_rule.go [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+	OrgID        int64             `json:"orgId"`
+	RuleGroup    string            `json:"ruleGroup"`
+	Title        string            `json:"title"`
+	UID          string            `json:"uid,omitempty"`
+	Updated      string            `json:"updated,omitempty"`
+	ForDuration  time.Duration     `json:"for"`
+	Provenance   string            `json:"provenance"`
+}
+
+type AlertQuery struct {
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion

// AlertQuery represents a single query stage associated with an alert definition.
type AlertQuery struct {
```

File: alerting_alert_rule.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+	path := fmt.Sprintf("/api/v1/provisioning/alert-rules/%s", uid)
+	result := AlertRule{}
+	err := c.request("GET", path, nil, nil, &result)
+	if err != nil {
+		return AlertRule{}, fmt.Errorf("alert rule with uid %s not found", uid)
+	}
+	return result, err
+}
+
+// NewAlertRule creates a new alert rule.
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
// NewAlertRule creates a new alert rule and returns its UID.
```

File: alerting_alert_rule.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+type ExecErrState string
+type NoDataState string
+
+const (
+	ErrOK       ExecErrState = "OK"
+	ErrError    ExecErrState = "Error"
+	ErrAlerting ExecErrState = "Alerting"
+	NoDataOk    NoDataState  = "OK"
+	NoData      NoDataState  = "NoData"
+	NoDataErr   NoDataState  = "Error"
```
Comment (alexweav) [inline alerting_alert_rule.go]:
```suggestion
	NoDataAlerting   NoDataState  = "Alerting"
```

File: alerting_alert_rule_test.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
+package gapi
+
+import (
+	"encoding/json"
+	"testing"
+
+	"github.com/gobs/pretty"
+)
+
+func TestAlertRules(t *testing.T) {
```
Comment (alexweav) [inline alerting_alert_rule_test.go:10]: Excellent tests. Thanks!

**Review by alexweav — CHANGES_REQUESTED** [review]
Just a few more points of feedback. These are all fairly nitpicky, this contribution is looking great!

No need to worry about the GET/PUT rule group endpoints as part of this PR quite yet. I'm actively working on those endpoints and will add them as soon as they're supported.

File: alerting_alert_rule.go (hunk 10 above)
Comment (alexweav) [reply alerting_alert_rule.go]: Based on: https://github.com/grafana/grafana/blob/e4bf7ddefe5aa315313f3c637a09cde0d7459937/pkg/services/ngalert/models/alert_rule.go#L48-L52

**Comment by stevenbrichards** [comment]
@alexweav thanks, applied all your suggestions, I'll rebase and squash

**Comment by stevenbrichards** [comment]
Good to go 👍

**Review by alexweav — APPROVED** [review]
Nearly there - I think the tests need a couple touch-ups based on the rename suggestions + another pass with the auto-formatter, but otherwise this PR is good to go. Thanks!

**Comment by stevenbrichards** [comment]
Well that's embarrassing 😅 fixed and go fmt'd. Thanks @alexweav

**Comment by stevenbrichards** [comment]
Tagged to re-review to merge

**Review by alexweav — APPROVED** [review]

**Review by JohnnyQQQQ — APPROVED** [review]

**Pull request merged**